### DISCUSSION TITLE
Cypress hangs when pulling binary

### DISCUSSION THREAD

**Discussion by ChrisJBurns** [cypress-io/cypress#27063] [CI setup] [answered]
I have run `npm ci`. There are no problems.

When I run `npx cypress version` I get the following:
```
Cypress package version: 12.12.0
Cypress binary version: not installed
Electron version: not found
Bundled Node version: not found
```

So I'm assuming the `postinstall` hook didn't work when installing cypress binary - not sure why it didn't fail in the `npm ci` step.

However, when I run `DEBUG="cypress:*" npx cypress install`, I get the following and it just hangs indefinitely:
```
root@080d34b8-fb36-4d26-5532-bf2606e78675:~# DEBUG="cypress:*" npx cypress install
  cypress:cli:cli cli starts with arguments ["/usr/local/nvm/versions/node/v16.20.0/bin/node","/tmp/build/e7b20f12/merge-request/service/node_modules/.bin/cypress","install"] +0ms
  cypress:cli NODE_OPTIONS is not set +0ms
  cypress:cli:cli program parsing arguments +1ms
  cypress:cli parsed cli options {} +160ms
  cypress:cli installing with options {} +0ms
  cypress:cli detecting arch { osPlatform: 'linux', osArch: 'x64' } +1ms
  cypress:cli arm uname -m result: { stdout: 'x86_64' }  +13ms
  cypress:cli version in package.json is 12.12.0, version to install is 12.12.0 +13ms
  cypress:cli Reading binary package.json from: /tmp/build/e7b20f12/merge-request/service/.cache/Cypress/12.12.0/Cypress/resources/app/package.json +0ms
  cypress:cli no binary installed under cli version +9ms
  cypress:cli checking local file /tmp/build/e7b20f12/12.12.0 cwd /tmp/build/e7b20f12/merge-request/service +0ms
  cypress:cli preparing to download and unzip version  12.12.0 to path /tmp/build/e7b20f12/merge-request/service/.cache/Cypress/12.12.0 +0ms
Installing Cypress (version: 12.12.0)


  cypress:cli needed Cypress version: 12.12.0 +0ms
  cypress:cli source url https://download.cypress.io/desktop/12.12.0?platform=linux&arch=x64 +0ms
  cypress:cli downloading cypress.zip to "/tmp/cypress-413.zip" +0ms
  cypress:cli Downloading package {
  url: 'https://download.cypress.io/desktop/12.12.0?platform=linux&arch=x64',
  proxy: null,
  downloadDestination: '/tmp/cypress-413.zip'
} +1ms
```

However, if I run ` curl -I "https://download.cypress.io/desktop/12.12.0?platform=linux&arch=x64" --output file.zip`, it downloads the zip file with no issue.

I have exec'd onto the CI container to debug further, but it seems that the binary isn't being installed by Cypress at all. It's just hanging

**Comment by antischematic** [comment]
Facing a similar problem, downloading the binary directly from https://download.cypress.io/desktop fails with checksum error. It fails when downloading over IPv4 but works over IPv6.

**Comment by MikeMcC399** [comment]
@ChrisJBurns 

To get more information you could use:

```
export DEBUG='cypress:*,extract-zip'
npm install cypress@12.12.0 --verbose --foreground-scripts
```

**Reply by ChrisJBurns** [reply, marked as answer]
Hi @MikeMcC399 

Can confirm after spending a few days debugging this on our end. The problem was in fact our Squid Proxy that we were using. I will post some information here for any other that may learn the trickiness of multi-az request routing :)

Because we currently have 2 Squid Proxy Instances that are randomly provisioned in 2 of the 3 Availability Zones in the London Region, the Network Load Balancer that we also provision to balance requests between the Squid Proxy instances creates Elastic Network Interfaces in all 3 AZ's. This means that sometimes when it receives requests and tries to route it to the appropriate ENI, the ENI that it routes the request to, is not in the same AZ as an existing Squid Proxy Instance. So we enabled the `enable_cross_zone_load_balancing` flag on the Squid Proxy Network Load Balancer Terraform resource so that it is able to route requests successfully so that it doesn't lose requests. We had to enable this because some builds were failing when running Cypress because when it tried to download and install Cypress during the `postinstall` it was getting routed through our Squid Proxy via `https_proxy` and getting lost in that AZ that didn't have a Squid Proxy instance. We enabled cross zone load balancing, and our builds are successfully passing consistently.

**Reply by MikeMcC399** [reply]
Hi @ChrisJBurns 

Well I wouldn't have guessed that was the reason for your problem and thanks for explaining! 🙂 

I was recently assisting another user attempting to install from a mirror, so I worked out for that how to get more debug information, which in the end allowed finding the root cause.

**Reply by pacion** [reply]
what is the fix?

**Reply by MikeMcC399** [reply]
@pacion

> what is the fix?

This was an individual network problem. It's unlikely you have exactly the same reason as the original poster. You could however use the same troubleshooting steps. It would be better for you to start a new discussion here or in the Cypress technical community on Discord

[![Discord chat](https://img.shields.io/badge/chat-on%20Discord-brightgreen)](https://on.cypress.io/discord) (click on button)
